pybind.obj : error LNK2001: 无法解析的外部符号 "class at::Tensor __cdecl nms_rotated(class at::Tensor,c lass at::Tensor,class at::Tensor,class at::Tensor,float,int)" (?nms_rotated@@YA?AVTensor@at@@V12@V12@11MH@Z)

OS: Win10
VS: 2019 (19.28.29335)
mmcv: 1.2.1
CUDA: 11.0
pytorch: 1.7.1
Python: 3.8.5 (tags/v3.8.5:580fbb0, Jul 20 2020, 15:57:54) [MSC v.1924 64 bit (AMD64)]

本人编译mmcv-full的环境介绍如上，但是出现了以下错误：
```
  正在创建库 E:\Downloads\mmcv-1.2.1\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\_ext.cp38
-win_amd64.lib 和对象 E:\Downloads\mmcv-1.2.1\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\
_ext.cp38-win_amd64.exp
pybind.obj : error LNK2001: 无法解析的外部符号 "class at::Tensor __cdecl nms_rotated(class at::Tensor,c
lass at::Tensor,class at::Tensor,class at::Tensor,float,int)" (?nms_rotated@@YA?AVTensor@at@@V12@V12@11MH@Z)
  已定义且可能匹配的符号上的提示:
    "class at::Tensor __cdecl nms_rotated(class at::Tensor,class at::Tensor,class at::Tensor,class at::Tensor,float,int)" (?nms_rotated@@YA?AVTensor@at@@V12@000MH@Z)
build\lib.win-amd64-3.8\mmcv\_ext.cp38-win_amd64.pyd : fatal error LNK1120: 1 个无法解析的外部命令
error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX86\\x64\\link.exe' failed with exit status 1120
```

看提示好像是link.exe的问题，奇怪的是编译的时候host是x86，不知道是不是这个问题，但是试了一下没有找到build_ext时手动设置msvc的host的方法，请问有人在Windows下成功编译了mmcv-full了吗？mmdetection依赖这个，否则用不了

I have tried but unfortunately the error still happened:
```
[40/40] C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\bin\nvcc --use-local-env -Xcompiler /MD -Xcompiler /wd4819 -Xcompiler /wd4251 -Xcompiler /wd4244 -Xcompiler /wd4267 -Xcompiler /wd4275 -Xcompiler /wd4018 -Xcompiler /wd4190 -Xcompiler /EHsc -Xcudafe --diag_suppress=base_class_has_different_dll_interface -Xcudafe --diag_suppress=field_without_dll_interface -Xcudafe --diag_suppress=dll_interface_conflict_none_assumed -Xcudafe --diag_suppress=dll_interface_conflict_dllexport_assumed -DMMCV_WITH_CUDA -ID:\Github\mmcv\mmcv\ops\csrc -ID:\Applications\WPy64-3850\python-3.8.5.amd64\lib\site-packages\torch\include -ID:\Applications\WPy64-3850\python-3.8.5.amd64\lib\site-packages\torch\include\torch\csrc\api\include -ID:\Applications\WPy64-3850\python-3.8.5.amd64\lib\site-packages\torch\include\TH -ID:\Applications\WPy64-3850\python-3.8.5.amd64\lib\site-packages\torch\include\THC "-IC:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\include" -ID:\Applications\WPy64-3850\python-3.8.5.amd64\include -ID:\Applications\WPy64-3850\python-3.8.5.amd64\include "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\include" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um" "-IE:\Windows Kits\10\include\10.0.18362.0\ucrt" "-IE:\Windows Kits\10\include\10.0.18362.0\shared" "-IE:\Windows Kits\10\include\10.0.18362.0\um" "-IE:\Windows Kits\10\include\10.0.18362.0\winrt" "-IE:\Windows Kits\10\include\10.0.18362.0\cppwinrt" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\include" -ID:\Applications\MKL\compilers_and_libraries_2020.4.311\windows\mkl\include "-IE:\Windows Kits\10\Include\10.0.18362.0\ucrt" "-IE:\Windows Kits\10\Include\10.0.18362.0\um" "-IE:\Windows Kits\10\Include\10.0.18362.0\shared" "-IE:\Windows Kits\10\Include\10.0.18362.0\winrt" -c D:\Github\mmcv\mmcv\ops\csrc\pytorch\tin_shift_cuda.cu -o D:\Github\mmcv\build\temp.win-amd64-3.8\Release\mmcv/ops/csrc/pytorch\tin_shift_cuda.obj -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_EXTENSION_NAME=_ext -D_GLIBCXX_USE_CXX11_ABI=0 -gencode=arch=compute_75,code=sm_75
tin_shift_cuda.cu
creating D:\Github\mmcv\build\lib.win-amd64-3.8
creating D:\Github\mmcv\build\lib.win-amd64-3.8\mmcv
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\bin\HostX86\x64\link.exe /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO /LIBPATH:D:\Applications\WPy64-3850\python-3.8.5.amd64\lib\site-packages\torch\lib "/LIBPATH:C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\lib/x64" /LIBPATH:D:\Applications\WPy64-3850\python-3.8.5.amd64\libs /LIBPATH:D:\Applications\WPy64-3850\python-3.8.5.amd64\PCbuild\amd64 "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\ATLMFC\lib\x64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\lib\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\lib\um\x64" "/LIBPATH:E:\Windows Kits\10\lib\10.0.18362.0\ucrt\x64" "/LIBPATH:E:\Windows Kits\10\lib\10.0.18362.0\um\x64" "/LIBPATH:E:\Windows Kits\10\Lib\10.0.18362.0\um\x64" "/LIBPATH:E:\Windows Kits\10\Lib\10.0.18362.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\lib\x64" /LIBPATH:D:\Applications\MKL\compilers_and_libraries_2020.4.311\windows\mkl\lib\intel64 /LIBPATH:D:\Applications\MKL\compilers_and_libraries_2020.4.311\windows\compiler\lib\intel64 c10.lib torch.lib torch_cpu.lib torch_python.lib cudart.lib c10_cuda.lib torch_cuda.lib /EXPORT:PyInit__ext D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\bbox_overlaps.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\bbox_overlaps_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\box_iou_rotated.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\box_iou_rotated_cpu.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\box_iou_rotated_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\carafe.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\carafe_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\carafe_naive.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\carafe_naive_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\cc_attention.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\cc_attention_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\corner_pool.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\deform_conv.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\deform_conv_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\deform_roi_pool.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\deform_roi_pool_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\focal_loss.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\focal_loss_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\info.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\masked_conv2d.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\masked_conv2d_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\modulated_deform_conv.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\modulated_deform_conv_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\nms.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\nms_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\nms_rotated.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\nms_rotated_cpu.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\nms_rotated_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\psamask.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\psamask_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\pybind.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\roi_align.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\roi_align_cpu.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\roi_align_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\roi_pool.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\roi_pool_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\sync_bn.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\sync_bn_cuda.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\tin_shift.obj D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\tin_shift_cuda.obj /OUT:build\lib.win-amd64-3.8\mmcv\_ext.cp38-win_amd64.pyd /IMPLIB:D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\_ext.cp38-win_amd64.lib
  正在创建库 D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\_ext.cp38-win_amd64.lib 和对象 D:\Github\mmcv\build\temp.win-amd64-3.8\Release\./mmcv/ops/csrc/pytorch\_ext.cp38-win_amd64.exp

pybind.obj : error LNK2001: 无法解析的外部符号 "class at::Tensor __cdecl nms_rotated(class at::Tensor,class at::Tensor,class at::Tensor,class at::Tensor,float,int)" (?nms_rotated@@YA?AVTensor@at@@V12@V12@11MH@Z)

  已定义且可能匹配的符号上的提示:
    "class at::Tensor __cdecl nms_rotated(class at::Tensor,class at::Tensor,class at::Tensor,class at::Tensor,float,int)" (?nms_rotated@@YA?AVTensor@at@@V12@000MH@Z)
build\lib.win-amd64-3.8\mmcv\_ext.cp38-win_amd64.pyd : fatal error LNK1120: 1 个无法解析的外部命令
error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX86\\x64\\link.exe' failed with exit status 1120
```

--- a/mmcv/ops/csrc/pytorch/pybind.cpp
+++ b/mmcv/ops/csrc/pytorch/pybind.cpp
@@ -178,8 +178,8 @@ Tensor top_pool_backward(Tensor input, Tensor grad_output);
 void box_iou_rotated(const Tensor boxes1, const Tensor boxes2, Tensor ious,
                      const bool aligned);
 
-Tensor nms_rotated(const Tensor dets, Tensor scores, Tensor order,
-                   Tensor dets_sorted, const float iou_threshold,
+Tensor nms_rotated(const Tensor dets, const Tensor scores, const Tensor order,
+                   const Tensor dets_sorted, const float iou_threshold,
                    const int multi_label);
 
 PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {

--- a/tests/test_ops/test_nms_rotated.py
+++ b/tests/test_ops/test_nms_rotated.py
@@ -1,12 +1,12 @@
 import numpy as np
 import torch
-
+import pytest
 
 class TestNmsRotated(object):
 
     def test_ml_nms_rotated(self):
         if not torch.cuda.is_available():
-            return
+            pytest.skip('test requires GPU')
         from mmcv.ops import nms_rotated
         np_boxes = np.array(
             [[6.0, 3.0, 8.0, 7.0, 0.5, 0.7], [3.0, 6.0, 9.0, 11.0, 0.6, 0.8],
@@ -31,7 +31,7 @@ class TestNmsRotated(object):
 
     def test_nms_rotated(self):
         if not torch.cuda.is_available():
-            return
+            pytest.skip('test requires GPU')
         from mmcv.ops import nms_rotated
         np_boxes = np.array(
             [[6.0, 3.0, 8.0, 7.0, 0.5, 0.7], [3.0, 6.0, 9.0, 11.0, 0.6, 0.8],

--- a/tests/test_ops/test_nms_rotated.py
+++ b/tests/test_ops/test_nms_rotated.py
@@ -3,11 +3,12 @@ import pytest
 import torch
 
 
-class TestNmsRotated(object):
+@pytest.mark.skipif(
+    not torch.cuda.is_available(),
+    reason='GPU is required to test NMSRotated op')
+class TestNmsRotated:
 
     def test_ml_nms_rotated(self):
-        if not torch.cuda.is_available():
-            pytest.skip('test requires GPU')
         from mmcv.ops import nms_rotated
         np_boxes = np.array(
             [[6.0, 3.0, 8.0, 7.0, 0.5, 0.7], [3.0, 6.0, 9.0, 11.0, 0.6, 0.8],
@@ -31,8 +32,6 @@ class TestNmsRotated(object):
         assert np.allclose(keep_inds.cpu().numpy(), np_expect_keep_inds)
 
     def test_nms_rotated(self):
-        if not torch.cuda.is_available():
-            pytest.skip('test requires GPU')
         from mmcv.ops import nms_rotated
         np_boxes = np.array(
             [[6.0, 3.0, 8.0, 7.0, 0.5, 0.7], [3.0, 6.0, 9.0, 11.0, 0.6, 0.8],

--- a/tests/test_ops/test_nms_rotated.py
+++ b/tests/test_ops/test_nms_rotated.py
@@ -1,6 +1,7 @@
 import numpy as np
-import torch
 import pytest
+import torch
+
 
 class TestNmsRotated(object):
 

--- a/tests/test_ops/test_nms_rotated.py
+++ b/tests/test_ops/test_nms_rotated.py
@@ -24,9 +24,9 @@ class TestNmsRotated(object):
         boxes = torch.from_numpy(np_boxes).cuda()
         labels = torch.from_numpy(np_labels).cuda()
 
-        dets, keep_inds = nms_rotated(boxes, 0.5, labels, True)
+        dets, keep_inds = nms_rotated(boxes[:, :5], boxes[:, -1], 0.5, labels)
 
-        assert np.allclose(dets.cpu().numpy(), np_expect_dets)
+        assert np.allclose(dets.cpu().numpy()[:, :5], np_expect_dets)
         assert np.allclose(keep_inds.cpu().numpy(), np_expect_keep_inds)
 
     def test_nms_rotated(self):
@@ -47,6 +47,6 @@ class TestNmsRotated(object):
 
         boxes = torch.from_numpy(np_boxes).cuda()
 
-        dets, keep_inds = nms_rotated(boxes, 0.5)
-        assert np.allclose(dets.cpu().numpy(), np_expect_dets)
+        dets, keep_inds = nms_rotated(boxes[:, :5], boxes[:, -1], 0.5)
+        assert np.allclose(dets.cpu().numpy()[:, :5], np_expect_dets)
         assert np.allclose(keep_inds.cpu().numpy(), np_expect_keep_inds)

